### PR TITLE
Added support for deleting directories asynchronously

### PR DESCRIPTION
**Pull request by snargledorf**

### CODE REVIEW

**Comment by snargledorf** [comment]
This addresses some of the methods mentioned in #1415

File: src/Renci.SshNet/Sftp/ISftpFile.cs [hunk 1]
```diff
@@ -227,6 +229,16 @@ public interface ISftpFile
...
         void Delete();
 
+        /// <summary>
+        /// Permanently deletes a file on remote machine.
+        /// </summary>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> to observe.</param>
+        /// <returns>A <see cref="Task"/> that represents the asynchronous delete operation.</returns>
+        /// <returns>
+        /// A <see cref="Task"/> reprisenting the delete operation.
+        /// </returns>
```
Comment (Rob-Hague) [inline src/Renci.SshNet/Sftp/ISftpFile.cs]:
```suggestion
        /// <summary>
        /// Permanently deletes a file on the remote machine.
        /// </summary>
        /// <param name="cancellationToken">The <see cref="CancellationToken"/> to observe.</param>
        /// <returns>A <see cref="Task"/> that represents the asynchronous delete operation.</returns>
```

File: src/Renci.SshNet/Sftp/SftpFile.cs [hunk 2]
```diff
@@ -468,6 +470,21 @@ public void Delete()
...
         }
 
+        /// <summary>
+        /// Asynchronosly requests to permanently delete a file on the remote machine.
+        /// </summary>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> to observe.</param>
+        /// <returns>A <see cref="Task"/> that represents the asynchronous delete operation.</returns>
+        /// <returns>
+        /// A <see cref="Task"/> reprisenting the delete operation.
+        /// </returns>
```
Comment (Rob-Hague) [inline src/Renci.SshNet/Sftp/SftpFile.cs]:
```suggestion
        /// <inheritdoc/>
```

File: src/Renci.SshNet/Sftp/SftpSession.cs [hunk 3]
```diff
@@ -1557,6 +1557,45 @@ public void RequestRmDir(string path)
             }
         }
 
+        /// <summary>
+        /// Asynchronously performs an SSH_FXP_RMDIR request.
+        /// </summary>
+        /// <param name="path">The path.</param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
+        /// <returns>A task that represents the asynchronous <c>SSH_FXP_RMDIR</c> request.</returns>
```
Comment (Rob-Hague) [inline src/Renci.SshNet/Sftp/SftpSession.cs]:
```suggestion
        /// <inheritdoc/>
```

File: src/Renci.SshNet/SftpClient.cs [hunk 4]
```diff
@@ -371,6 +371,35 @@ public void DeleteDirectory(string path)
             _sftpSession.RequestRmDir(fullPath);
         }
 
+        /// <summary>
+        /// Asynchronously deletes remote directory specified by path.
+        /// </summary>
+        /// <param name="path">Directory to be deleted path.</param>
```
Comment (Rob-Hague) [inline src/Renci.SshNet/SftpClient.cs]:
```suggestion
        /// Asynchronously deletes a remote directory.
        /// </summary>
        /// <param name="path">The path of the directory to be deleted.</param>
```

File: src/Renci.SshNet/SftpClient.cs [hunk 5]
```diff
@@ -371,6 +371,35 @@ public void DeleteDirectory(string path)
...
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> to observe.</param>
+        /// <returns>A <see cref="Task"/> that represents the asynchronous delete operation.</returns>
+        /// <exception cref="ArgumentException"><paramref name="path"/> is <see langword="null"/> or contains only whitespace characters.</exception>
+        /// <exception cref="SshConnectionException">Client is not connected.</exception>
+        /// <exception cref="SftpPathNotFoundException"><paramref name="path"/> was not found on the remote host.</exception>
+        /// <exception cref="SftpPermissionDeniedException">Permission to delete the directory was denied by the remote host. <para>-or-</para> A SSH command was denied by the server.</exception>
+        /// <exception cref="SshException">A SSH error where <see cref="Exception.Message"/> is the message from the remote host.</exception>
+        /// <exception cref="ObjectDisposedException">The method was called after the client was disposed.</exception>
+        public async Task DeleteDirectoryAsync(string path, CancellationToken cancellationToken = default)
```
Comment (Rob-Hague) [inline src/Renci.SshNet/SftpClient.cs:428]: Can you also add it to `ISftpClient`? People like to use it for mocking

File: src/Renci.SshNet/SftpClient.cs [hunk 6]
```diff
@@ -1474,6 +1503,21 @@ public void Delete(string path)
...
 
+        /// <summary>
+        /// Permanently deletes a file on remote machine.
+        /// </summary>
+        /// <param name="path">The name of the file or directory to be deleted. Wildcard characters are not supported.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> to observe.</param>
+        /// <returns>A <see cref="Task"/> that represents the asynchronous delete operation.</returns>
+        /// <returns>
+        /// A <see cref="Task"/> reprisenting the delete operation.
+        /// </returns>
```
Comment (Rob-Hague) [inline src/Renci.SshNet/SftpClient.cs]:
```suggestion
        /// <summary>
        /// Permanently deletes a file on the remote machine.
        /// </summary>
        /// <param name="path">The name of the file or directory to be deleted. Wildcard characters are not supported.</param>
        /// <param name="cancellationToken">The <see cref="CancellationToken"/> to observe.</param>
        /// <returns>A <see cref="Task"/> that represents the asynchronous delete operation.</returns>
```

And add to the interface

File: test/Renci.SshNet.IntegrationTests/SftpClientTests.cs [hunk 7]
```diff
@@ -83,8 +83,8 @@ public async Task Create_directory_with_contents_and_list_it_async()
                 actualFiles.Add((file.FullName, file.IsRegularFile, file.IsDirectory));
             }
 
-            _sftpClient.DeleteFile(testFilePath);
-            _sftpClient.DeleteDirectory(testDirectory);
+            await _sftpClient.DeleteFileAsync(testFilePath, CancellationToken.None);
+            await _sftpClient.DeleteDirectoryAsync(testDirectory, CancellationToken.None);
```
Comment (Rob-Hague) [inline test/Renci.SshNet.IntegrationTests/SftpClientTests.cs:87]: I know you are mirroring the sync version, but I don't think it's really testing deletion. If not adding a separate test, perhaps at least another call to ListDirectory/Exists here? e.g.

 ```suggestion
            await _sftpClient.DeleteFileAsync(testFilePath, CancellationToken.None);
            
            // check the file no longer exists
            // ...
            
            await _sftpClient.DeleteDirectoryAsync(testDirectory, CancellationToken.None);
            
            // check the directory no longer exists
            // ...
```

Comment (snargledorf) [reply test/Renci.SshNet.IntegrationTests/SftpClientTests.cs:87]: Yes agreed, this was a bit lazy on my part. I will work on a new test to verify async delete functionality.

**Comment by snargledorf** [comment]
Added additional tests and added requested method declarations to ISftpClient.

**Review by Rob-Hague — APPROVED** [review]
Thanks

**Pull request merged**